src/protocols/sfp.stanza

```diff
@@ -0,0 +1,334 @@
+<DOC>
+
+public defn SFP-get-skew-loss-vals (link:jsl/protocols/sfp/SFPLink) -> [Toleranced, Toleranced, Double]:
+  ; skew, lane-skew, loss
+  switch(link) :
+    link-SFP       : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-cSFP      : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-SFP+      : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-SFP28     : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-SFP56     : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-SFP-DD    : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-SFP112    : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-SFP-DD112 : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-QSFP      : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-QSFP+     : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-QSFP28    : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-QSFP56    : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-QSFP112   : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+    link-QSFP-DD   : [0.0 +/- 1.0e-12, 0.0 +/- 100.0e-12, 5.0]
+
+doc: \<DOC>
+@brief Differential impedance specified by the SFP standard
+This is a helper function that returns the expected differential
+trace impedance.
+
+@return Upper/lower limits for the impedance.
+<DOC>
+
+public defn SFP-get-trace-impedance () -> Toleranced :
+  100.0 +/- (10 %)
+
+doc: \<DOC>
+SFP Constraint
+This constraint is used to constrain lane(s) of the SFP interface.
+This type would typically be used with a {@link type:LaneConstraint}
+type for constructing a full-duplex SFP interface.
+This constraint expects to be passed `sfp` bundle ports to constrain.
+
+<DOC>
+public defstruct SFP-Constraint <: SI-Constraint :
+  doc : \<DOC>
+  SFP Specification
+  <DOC>
+  link : jsl/protocols/sfp/SFPLink
+  doc: \<DOC>
+  Specification for the intra-pair skew
+  <DOC>
+  skew : Toleranced
+  doc: \<DOC>
+  Max skew between all of the lanes in a single (Q)SFP(-DD) link
+  <DOC>
+  lane-skew : Toleranced
+  doc: \<DOC>
+  Max Loss Limit in dB
+  <DOC>
+  loss : Double with:
+    ensure => ensure-positive!
+  doc: \<DOC>
+  Differential routing structure for all signals.
+  <DOC>
+  route-struct : DifferentialRoutingStructure
+with:
+  keyword-constructor => true
+  constructor => #SFP-Constraint
+
+doc: \<DOC>
+Shorthand constructor for the SFP Constraint
+
+@param link The SFP link specification
+@param route-struct The differential routing structure
+<DOC>
+public defn SFP-Constraint (
+  --
+  link : jsl/protocols/sfp/SFPLink
+  route-struct : DifferentialRoutingStructure
+  ) -> SFP-Constraint :
+  val [ps, ls, ml] = SFP-get-skew-loss-vals(link)
+  SFP-Constraint(
+    link = link,
+    skew = ps,
+    lane-skew = ls,
+    loss = ml,
+    route-struct = route-struct
+  )
+
+defn route-constrain (cst:SFP-Constraint, route:KeyValue<JITXObject, JITXObject>) :
+  skew-window(skew(cst), route)
+  max-loss(loss(cst), route)
+  diff-structure(route-struct(cst), route)
```

Comment:
I think this is just a duplicate of the `DiffPair-Constraint` - you should be able to reuse `diffpair-constrain` if you use the appropriate type (ie, derive SFP-Constraint from `DiffPair-Constraint`